### PR TITLE
Fix saving meta when using `save_post`.

### LINKED ISSUE #13960: [3.0] Saving product meta on `save_post` no longer works
Extensions that save product meta on `save_post` (priority > 1), such as Bookings and Subscriptions are failing to update their meta when `WC_Data::maybe_read_meta_data` is called on the `woocommerce_admin_process_product_object` action by another plugin/extension.

It's not yet entirely clear to me why, but this is causing conflicts with all extensions that have been updated for 3.0 and save their meta on `woocommerce_admin_process_product_object` using the `WC_Data` methods.

Might be worth giving some attention to this as it looks serious. @thenbrent @mikejolley @claudiulodro

### PR DESCRIPTION
**Pull request by emdashcodes**
Fixes #13960. Here is the general flow of the bug:

* `WC_Meta_Box_Product_Data::save()` calls `woocommerce_admin_process_product_object` and passes the `$product` object in.
* NYP (or another plugin) hooks to `woocommerce_admin_process_product_object` and sets some meta with `$product->update_meta_data()`. 
* `WC_Meta_Box_Product_Data::save()`  calls `$product->save()`. Our meta is currently unsaved in the database at this time, but is present on the `$product` object.  
* Down the chain from `$product->save()` is the product data store `update` method, which calls  `wp_update_post` and runs certain hooks like `save_post`.
* Subscriptions (or another plugin) hooks in `save_post` and does direct `update_post_meta` calls.
* Back in the product data store, after `wp_update_post`, we run CRUD `update_post_meta` and `save_post_meta`. The data store and CRUD object don't know about the direct meta changes and `save_meta_data` wipe out the changes thinking the value Is still the old value (`save_meta_data` loops through an array of meta and calls `data_store->update_meta` -- so the old value is getting set here).

Tl;dr Order of operations + a mix of CRUD and direct meta updates causes issues.

--

To fix, I've moved `save_meta_data` earlier in the update routine and then re pull the meta cache after `save_post` gets called. I've done this for the data stores that use those internal WP functions.

Our meta handling could probably be made smarter -- but I think that requires a bit more thought, especially when direct calls can still be made.

--

To Test:

* Make sure `phpunit` tests all pass.
* Install Name Your Price and Subscriptions. Activate both.
* On a subscriptions product, try to updating a value like subscription trial length and the NYP price fields. Make sure both save (without this branch, only NYP will save).
* Spot check updating coupons, products, and an order detail since there are some similar changes there as well.

### CODE REVIEW

**Comment by franticpsyx** [comment]
@justinshreve -- The whole flow you described is itself hooked into `save_post`. I could be wrong but have you considered the implications of having the `save_post` (in the 4th step above) action fired from within itself?

1.

Have a look at this -- https://core.trac.wordpress.org/ticket/17817 

With the flow you describe, under WP 4.6 I would expect any action hooked into `save_post` after [priority 1](https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-meta-boxes.php#L43) to fail firing. 

2.

https://codex.wordpress.org/Function_Reference/wp_update_post#Caution_-_Infinite_loop

(2) should have been discovered if there was a real issue, but I'm still sharing my thoughts in case a plugin is affected.

Of course these are different issues than the ones we are talking about here. (1) could be serious enough to:

- document somewhere; and/or
- consider a different approach.

e.g. perhaps replace `wp_update_post` with something more direct?

**Comment by franticpsyx** [comment]
Also the process above doesn't make it clear to me why WCS would fail writing its own meta on the "outer" `save_post` that this flow is hooked into.

**Comment by justinshreve** [comment]
From what I can tell, [this check](https://github.com/woocommerce/woocommerce/blob/14802a03079f4725c4c75d53c7bca1a22d5be7df/includes/admin/class-wc-admin-meta-boxes.php#L186) helps prevent against those loops. The initial `save_post` is only called once early. Seems like the important thing to document here would be to make sure everything is higher priority than 1. 

I just tested on `4.6.4` and have not run into any issues with this PR. 

> Also the process above doesn't make it clear to me why WCS would fail writing its own meta on the "outer" save_post that this flow is hooked into.

It doesn't fail to write. It gets written and then looks like it gets wiped out in `save_meta_data`.

cc @mikejolley for review since it looks like you've dealt with these loops before.

**Comment by mikejolley** [comment]
> Also the process above doesn't make it clear to me why WCS would fail writing its own meta on the "outer" save_post that this flow is hooked into.

Since WooCommerce has always used save_post in admin, and used wp_update_post, the functions it hooks into save_post unhook themselves to prevent infinite loops. Thats why they do not fail here either.

@justinshreve @franticpsyx 

This could be avoided with direct wpdb calls and a clean post cache. It could even be faster. It will then not trigger save_post action. I think thats separate though.

As for the PR, it's a shame we have to do the read as it will slow things down. The alternative would be to track changes to meta like we do props but that may be more difficult to implement. OR would that not fix the issue either?

**Comment by justinshreve** [comment]
> This could be avoided with direct wpdb calls and a clean post cache. It could even be faster. It will then not trigger save_post action. I think thats separate though.

That would also be a bigger backwards incompatible change IMO since things are already hooking in there, so definitely not for a point release.

> As for the PR, it's a shame we have to do the read as it will slow things down. The alternative would be to track changes to meta like we do props but that may be more difficult to implement. OR would that not fix the issue either?

I think that might be part of making the meta handling smarter and could be another fix for this, but I do think it needs some more thought. I think there would still be an issue where things added directly with WP's `update_post_meta`might not still be picked up without re-reading so we would need to figure something out.

Since this bug is currently breaking extension management though, I think we could use this as a temporary fix (assuming it looks okay otherwise) and open an issue for smarter handling.

**Comment by mikejolley** [comment]
Seems to work and tests pass.
